### PR TITLE
modify  predict/predictClass function 

### PR DESCRIPTION
**Pull request by lingxiao1989**
## What changes were proposed in this pull request?
Added predictBeta and predictClassBeta function, which return sample and predict label together.

## How was this patch tested?
Just for code review
(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If it is possible, please attach a screenshot; otherwise, remove this)

## Related links or issues (optional)
Jenkins link: http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/841/

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala [hunk 1]
```diff
@@ -151,6 +151,40 @@ object Predictor {
...
+    val otherBroad = rdd.sparkContext.broadcast(SampleToMiniBatch(
+      batchSize = totalBatch,
+      partitionNum = Some(rdd.partitions.length),
+      featurePaddingParam = featurePaddingParam))
+    val batchOut = rdd.mapPartitions { partition =>
+      val localModel = modelBroad.value()
+      val localTransformer = otherBroad.value.cloneTransformer()
+      val miniBatch = localTransformer(partition)
+      miniBatch.flatMap(batch => {
+        val output = localModel.forward(batch.getInput)
```
Comment (i8run) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: It seems some duplicated code here.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 2]
```diff
@@ -640,6 +640,12 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
     Predictor(this).predict(dataset, batchSize, shareBuffer)
   }
 
+  final def predictBeta(dataset: RDD[Sample[T]],
```
Comment (i8run) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: It should be the same method name with different arguments.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala [hunk 3]
```diff
@@ -111,6 +112,15 @@ abstract class Sample[T: ClassTag] extends Serializable {
     throw new UnsupportedOperationException("Sample.set(): unimplemented deprecated method")
   }
 
+  /**
+   * Set data of label.
+   * @param labelData
+   * @return
+   */
+  def setLabel(labelData: Activity)(implicit ev: TensorNumeric[T]): Sample[T] = {
+    throw new UnsupportedOperationException("Sample.setLabel(): unimplemented deprecated method")
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala]: why deprecated ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 4]
```diff
@@ -634,9 +634,11 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
    *                  if -1, default is 4 * partitionNumber of datatset
    * @param shareBuffer whether to share same memory for each batch predict results
    */
+
+
   final def predict(dataset: RDD[Sample[T]],
     batchSize: Int = -1,
-    shareBuffer: Boolean = false): RDD[Activity] = {
+    shareBuffer: Boolean = false): RDD[Sample[T]] = {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: why change this ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala [hunk 5]
```diff
@@ -153,7 +153,7 @@ object Predictor {
 
   def predict[T: ClassTag](dataSet: RDD[Sample[T]], batchSize: Int = -1,
     shareBuffer: Boolean = false, model: Module[T], batchPerPartition: Int,
-    featurePaddingParam: Option[PaddingParam[T]])(implicit ev: TensorNumeric[T]): RDD[Activity] = {
+    featurePaddingParam: Option[PaddingParam[T]])(implicit ev: TensorNumeric[T]): RDD[Sample[T]] = {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: why change this ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala [hunk 6]
```diff
@@ -111,6 +112,22 @@ abstract class Sample[T: ClassTag] extends Serializable {
     throw new UnsupportedOperationException("Sample.set(): unimplemented deprecated method")
   }
 
+  /**
+   * Set data of label.
+   * @param labelData
+   * @return
+   */
+  def setLabel(labelData: Activity)(implicit ev: TensorNumeric[T]): Unit = {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala]: just name it `label`

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala [hunk 7]
```diff
@@ -111,6 +112,22 @@ abstract class Sample[T: ClassTag] extends Serializable {
...
+    throw new UnsupportedOperationException("Sample.setLabel(): unimplemented method")
+  }
+  /**
+    * Get data of label.
+    * @return label
+    */
+  def getLabel(): Array[Int] = {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala]: why `setLabel` parameter is activity and getLabel returns array of Int ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala [hunk 8]
```diff
@@ -180,6 +180,41 @@ object Predictor {
     }
   }
 
+  def predictClassTest[T: ClassTag](dataSet: RDD[Sample[T]], batchSize: Int = -1, model: Module[T],
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: why define a testing class here ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala [hunk 9]
```diff
@@ -2006,7 +2006,10 @@ class PythonBigDL[T: ClassTag](implicit ev: TensorNumeric[T]) extends Serializab
   def modelPredictClass(model: AbstractModule[Activity, Activity, T],
                       dataRdd: JavaRDD[Sample]): JavaRDD[Int] = {
     val sampleRdd = toJSample(dataRdd)
-    val tensorRDD = model.predictClass(sampleRdd)
+    val tensorRDD = model.predictClass(sampleRdd).map(sample => {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala]: why Python signature not changed ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 10]
```diff
@@ -646,8 +646,12 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
    * @param batchSize total batchSize for all partitions.
    *                  if -1, default is 4 * partitionNumber of dataset
    */
+
   final def predictClass(dataset: RDD[Sample[T]], batchSize: Int = -1): RDD[Int] = {
-    Predictor(this).predictClass(dataset, batchSize)
+    Predictor(this).predictClass(dataset, batchSize).map(sample => {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: why this is still RDD[Int] ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala [hunk 11]
```diff
@@ -180,6 +180,41 @@ object Predictor {
...
+    val batchOut = rdd.mapPartitions { partition =>
+      val localModel = modelBroad.value()
+      val localTransformer = otherBroad.value.cloneTransformer()
+      val miniBatch = localTransformer(partition)
+      miniBatch.flatMap(batch => {
+        val output = localModel.forward(batch.getInput)
+        splitBatch(output, shareBuffer, batch.size())
+      })
+    }
+    dataSet.zip(batchOut).map(sample => {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: How can we make sure the order is consistent ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala [hunk 12]
```diff
@@ -202,7 +203,12 @@ class ArraySample[T: ClassTag] private[bigdl](
       null
     }
   }
-
+/*
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala]: please remove this

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala [hunk 13]
```diff
@@ -2009,6 +2009,16 @@ class PythonBigDL[T: ClassTag](implicit ev: TensorNumeric[T]) extends Serializab
     model.predictClass(sampleRdd).map(Sample => Sample[T]).toJavaRDD()
   }
 
+  def getSampleLabel(model: AbstractModule[Activity, Activity, T],
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala]: it does not make sense to expose this method, it's out of order

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala [hunk 14]
```diff
@@ -2004,9 +2004,18 @@ class PythonBigDL[T: ClassTag](implicit ev: TensorNumeric[T]) extends Serializab
...
   def modelPredictClass(model: AbstractModule[Activity, Activity, T],
-                      dataRdd: JavaRDD[Sample]): JavaRDD[Int] = {
+                      dataRdd: JavaRDD[Sample]): JavaRDD[Sample] = {
     val sampleRdd = toJSample(dataRdd)
-    val tensorRDD = model.predictClass(sampleRdd)
+    val tensorRDD = model.predictClass(sampleRdd).map(toPySample(_))
+    new JavaRDD[Sample](tensorRDD)
+  }
+
+  def getSampleLabel(dataRdd: JavaRDD[Sample]): JavaRDD[Int] = {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala]: it does not make any sense to return a RDD if Int

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala [hunk 15]
```diff
@@ -2004,10 +2004,19 @@ class PythonBigDL[T: ClassTag](implicit ev: TensorNumeric[T]) extends Serializab
...
-                      dataRdd: JavaRDD[Sample]): JavaRDD[Int] = {
+                      dataRdd: JavaRDD[Sample]): JavaRDD[Sample] = {
     val sampleRdd = toJSample(dataRdd)
-    val tensorRDD = model.predictClass(sampleRdd)
-    new JavaRDD[Int](tensorRDD)
+    val tensorRDD = model.predictClass(sampleRdd).map(toPySample(_))
+    new JavaRDD[Sample](tensorRDD)
+  }
+
+  def getSampleLabel(dataRdd: JavaRDD[Sample]): JList[Int] = {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala]: why still returning a list for RDD input ?

File: pyspark/test/bigdl/test_simple_integration.py [hunk 16]
```diff
@@ -544,9 +545,14 @@ def test_predict(self):
             assert_allclose(p_with_batch[i], ground_label[i], atol=1e-6, rtol=0)
 
         predict_class = model.predict_class(predict_data)
-        predict_labels = predict_class.take(6)
-        for i in range(0, total_length):
-            assert predict_labels[i] == 1
+        if isinstance(predict_class, RDD):
+            for sample in predict_class:
```
Comment (wzhongyuan) [inline pyspark/test/bigdl/test_simple_integration.py]: I am confusing why it's working. if it's RDD, we should execute action first before iterate through it, no ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala (hunk 12 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala]: Remove unused code

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala (hunk 11 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: you need to resolve it

File: pyspark/test/bigdl/test_simple_integration.py (hunk 16 above)
Comment (lingxiao1989) [reply pyspark/test/bigdl/test_simple_integration.py]: I searched ways to iterate a RDD on the Internet.. I think I could follow this example. I will change my code. 

# create Spark context with Spark configuration
  conf = SparkConf().setAppName("Print Contents of RDD - Python")
  sc = SparkContext(conf=conf)
 
  # read input text file to RDD
  rdd = sc.textFile("data/rdd/input/file1.txt")
 
  # collect the RDD to a list
  list_elements = rdd.collect()
 
  # print the list
  for element in list_elements:
    print(element)

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala (hunk 11 above)
Comment (lingxiao1989) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: Ok. I am fixing it. Will move the zip step into each map partitions.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala [hunk 17]
```diff
@@ -194,6 +228,7 @@ object Predictor {
       })
     }
   }
+*/
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: why leave it commented ?

Comment (lingxiao1989) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: That is the old predictClass api. Should I remove it or keep it with a @deprecated annotation?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala (hunk 12 above)
Comment (lingxiao1989) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala]: I will remove it

File: pyspark/test/bigdl/test_simple_integration.py [hunk 18]
```diff
@@ -544,9 +545,14 @@ def test_predict(self):
             assert_allclose(p_with_batch[i], ground_label[i], atol=1e-6, rtol=0)
 
         predict_class = model.predict_class(predict_data)
-        predict_labels = predict_class.take(6)
-        for i in range(0, total_length):
-            assert predict_labels[i] == 1
+        if isinstance(predict_class, RDD):
```
Comment (wzhongyuan) [inline pyspark/test/bigdl/test_simple_integration.py]: why we need to check if it's RDD ?  it will only return RDD right ?

Comment (lingxiao1989) [reply pyspark/test/bigdl/test_simple_integration.py]: In local mode, predict_class calls the predictLocalClass API and get a JList[Int] result.
and in distributed mode, predict_class calls the modelPredictClass API and gets a JavaRDD[Sample]. That why I use the if statement here and handle the result separately.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala [hunk 19]
```diff
@@ -2004,10 +2004,10 @@ class PythonBigDL[T: ClassTag](implicit ev: TensorNumeric[T]) extends Serializab
   }
 
   def modelPredictClass(model: AbstractModule[Activity, Activity, T],
-                      dataRdd: JavaRDD[Sample]): JavaRDD[Int] = {
+                      dataRdd: JavaRDD[Sample]): JavaRDD[Sample] = {
     val sampleRdd = toJSample(dataRdd)
-    val tensorRDD = model.predictClass(sampleRdd)
-    new JavaRDD[Int](tensorRDD)
+    val tensorRDD = model.predictClass(sampleRdd).map(toPySample(_))
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala]: name it sampleRDD

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala [hunk 20]
```diff
@@ -209,7 +209,11 @@ object Predictor {
         splitBatch(output, shareBuffer, batch.size())
       })
       partition.zip(batchOut).map(sample => {
-        Sample(sample._1.feature(), sample._2.toTensor)
+        if (sample._2.toTensor == null) {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: why this change ? isn't it supposed to have label after prediction ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala [hunk 21]
```diff
@@ -2005,8 +2005,8 @@ class PythonBigDL[T: ClassTag](implicit ev: TensorNumeric[T]) extends Serializab
 
   def modelPredictClass(model: AbstractModule[Activity, Activity, T],
                       dataRdd: JavaRDD[Sample]): JavaRDD[Sample] = {
-    val sampleRdd = toJSample(dataRdd)
-    val tensorRDD = model.predictClass(sampleRdd).map(toPySample(_))
+    val sampleRDD = toJSample(dataRdd)
+    val tensorRDD = model.predictClass(sampleRDD).map(toPySample(_))
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala]: pySampleRDD ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala (hunk 20 above)
Comment (lingxiao1989) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: I think the batchOut are the predicted label array. So I zip the partition with the batchOut.

Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: I mean why you add the null checking ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala [hunk 22]
```diff
@@ -182,15 +180,38 @@ object Predictor {
...
+    val localBatchPerPartition = totalBatch / realPartitionLength
+    rdd.mapPartitions { partition =>
+      val localModel = modelBroad.value()
+      val localTransformer = otherBroad.value.cloneTransformer()
+      val miniBatch = localTransformer(partition)
+      val batchOut = miniBatch.flatMap(batch => {
+        val output = localModel.forward(batch.getInput)
+        splitBatch(output, shareBuffer, batch.size())
+      })
+      partition.grouped(localBatchPerPartition).flatMap(samples => {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Predictor.scala]: I did not see where the `batchOut` is used

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala [hunk 23]
```diff
@@ -2004,10 +2004,10 @@ class PythonBigDL[T: ClassTag](implicit ev: TensorNumeric[T]) extends Serializab
   }
 
   def modelPredictClass(model: AbstractModule[Activity, Activity, T],
-                      dataRdd: JavaRDD[Sample]): JavaRDD[Int] = {
-    val sampleRdd = toJSample(dataRdd)
-    val tensorRDD = model.predictClass(sampleRdd)
-    new JavaRDD[Int](tensorRDD)
+                      dataRdd: JavaRDD[Sample]): JavaRDD[Sample] = {
+    val sampleRDD = toJSample(dataRdd)
+    val PysampleRDD = model.predictClass(sampleRDD).map(toPySample(_))
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala]: lower case for p

**Comment by lingxiao1989** [comment]
Jenkins link: http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/841/

**Comment by wzhongyuan** [comment]
LGTM
